Sources/Hummingbird/ApplicationBuilder.swift

```diff
@@ -0,0 +1,140 @@
+/// run separately with `buildAndRun`.
+/// ```
+/// let app = HBApplicationBuilder()
+/// app.middleware.add(MyMiddleware())
+/// app.router.get("hello") { _ in
+///     return "hello"
+/// }
+/// try await app.buildAndRun()
+/// ```
+/// Editing the application builder setup after calling `build` will produce undefined behaviour.
+public final class HBApplicationBuilder {
+    // MARK: Member variables
+
+    /// event loop group used by application
+    public let eventLoopGroup: EventLoopGroup
+    /// thread pool used by application
+    public let threadPool: NIOThreadPool
+    /// routes requests to requestResponders based on URI
+    public let router: HBRouterBuilder
+    /// Configuration
+    public let configuration: HBApplication.Configuration
+    /// Logger. Required to be a var by hummingbird-lambda
+    public var logger: Logger
+    /// Encoder used by router
+    public var encoder: HBResponseEncoder
+    /// decoder used by router
+    public var decoder: HBRequestDecoder
+    /// on server running
+    public var onServerRunning: @Sendable (Channel) async -> Void
+    /// additional channel handlers
+    var additionalChannelHandlers: [@Sendable () -> any RemovableChannelHandler]
+
+    // MARK: Initialization
+
+    /// Initialize new Application
+    public init(
+        configuration: HBApplication.Configuration = HBApplication.Configuration(),
+        eventLoopGroupProvider: EventLoopGroupProvider = .global
+    ) {
+        var logger = Logger(label: configuration.serverName ?? "HummingBird")
+        logger.logLevel = configuration.logLevel
+        self.logger = logger
+
+        self.router = HBRouterBuilder()
+        self.configuration = configuration
+        self.encoder = NullEncoder()
+        self.decoder = NullDecoder()
+        self.onServerRunning = { _ in }
+        // add idle read, write handlers
+        if let idleTimeoutConfiguration = configuration.idleTimeoutConfiguration {
+            self.additionalChannelHandlers = [{
+                IdleStateHandler(
+                    readTimeout: idleTimeoutConfiguration.readTimeout,
+                    writeTimeout: idleTimeoutConfiguration.writeTimeout
+                )
+            }]
+        } else {
+            self.additionalChannelHandlers = []
+        }
+
+        // create eventLoopGroup
+        switch eventLoopGroupProvider {
+        case .global:
+            #if os(iOS)
+            self.eventLoopGroup = NIOTSEventLoopGroup.singleton
+            #else
+            self.eventLoopGroup = MultiThreadedEventLoopGroup.singleton
+            #endif
+        case .shared(let elg):
+            self.eventLoopGroup = elg
+        }
+
+        self.threadPool = NIOThreadPool(numberOfThreads: configuration.threadPoolSize)
+        self.threadPool.start()
+    }
+
+    // MARK: Methods
+
+    __consuming public func build() -> HBApplication {
```

Comment:
Do we expect `__consuming` to stably stick around in/after 5.9?

Reply:
The stdlib is full of functions labelled `__consuming`. I added it because I was thinking of making the builder a move only type, so once it had built the application it was consumed and unavailable

Reply:
That makes sense to me, and I really like it. I'm just worried that it might not be supported down the line.